### ISSUE TITLE
MetadataTypeAttribute doesn't influence DataAnnotations validation result

### ISSUE THREAD

**Issue by RichardBr** [dotnet/runtime#46678]
### Update by @SteveSandersonMS: please see the comment https://github.com/dotnet/runtime/issues/46678#issuecomment-747533844 for a description of the underlying issue/question.

Please ignore the parts of this thread that are specific to Blazor. That's just how the issue was originally reported, but it looks like the underlying concern is about DataAnnotations itself.

---

### Describe the bug
When I use the MetadataType attribute (from System.ComponentModel.DataAnnotations) on the class for the purpose of validation I get the following error in the console window
`WASM: System.TypeLoadException: Could not resolve type with token 01000027 from typeref (expected class 'System.ComponentModel.DataAnnotations.MetadataTypeAttribute' in assembly 'System.ComponentModel.Annotations, Version=4.3.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a')`

Greater details of the error are shown in this file...[console screenshot - validation error](https://user-images.githubusercontent.com/5938313/79199691-42595400-7e2d-11ea-95bc-3c2248d8b60e.png

Please note my sample application does works when I put all validation attributes into main the class.

For clarity:
This approach works...

```
public partial class PersonViewModel
{
    [Required(ErrorMessage = "First name is required.")]
    public string Fname { get; set; }
}
```

This approach fails...

```
public partial class PersonViewModel
{
    public string Fname { get; set; }
}

[MetadataType(typeof(PersonViewModelMetaData))]
public partial class PersonViewModel
{
}

public class PersonViewModelMetaData
{
    (ErrorMessage = "First name is required.")]
    public string Fname { get; set; }
}
```

### To Reproduce
Please run the attached sample blazor application. To see the error please click the navigation link labelled "validation - failure". On the page, just simply press the submit button to see the error.
[BlazorAppValidation.zip](https://github.com/dotnet/aspnetcore/files/4474042/BlazorAppValidation.zip)

### Further technical details
- Include the output of `dotnet --info`
```
.NET Core SDK (reflecting any global.json):
 Version:   3.1.300-preview-015048
 Commit:    13f19b4682

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.300-preview-015048\

Host (useful for support):
  Version: 3.1.3
  Commit:  4a9f85e9f8

.NET Core SDKs installed:
  3.1.101 [C:\Program Files\dotnet\sdk]
  3.1.201 [C:\Program Files\dotnet\sdk]
  3.1.300-preview-015048 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download

```

**Comment by julior77** [comment]
Any news on this issue?
As mentioned this is really important for teams using Scaffold-DbContext.

Any alternatives as a short term?

**Comment by DiegoVenancioVieira** [comment]
> Scaffold-DbContext
> Any news on this issue?
> As mentioned this is really important for teams using Scaffold-DbContext.
> Any alternatives as a short term?

No

**Comment by Xaeco** [comment]
Did major work migrating my project to Blazor and didn't see this one coming.
Now I'm stuck. 
Any workarounds would be appreciated.

**Comment by navidzehi** [comment]
@Xaeco  Maybe this can help you for now [C# new source generator](https://www.youtube.com/watch?v=3YwwdoRg2F4)

**Comment by DiegoVenancioVieira** [comment]
> @Xaeco Maybe this can help you for now [C# new source generator](https://www.youtube.com/watch?v=3YwwdoRg2F4)

@navidzehi Can you tell us HOW? Which part of video ?

**Comment by navidzehi** [comment]
@DiegoVenancioVieira well,I haven't try it myself but I think it's possible to do the following.
 You scaffold or manually create your model classes normally :

```
[MetadataType(typeof(PersonMetaData))]
public partial class Person
{
   public int Id {get;set;}
   public string FirstName{get;set;}
   public string LastName{get;set;}
}
```
then you may also have the following classes :
```
// in another partial file :
public partial class Person
{
 [NoMap]
  public string FullName => FirstName+" "+LastName;
}

public class PersonMetaData
{
   public object Id {get;set;}
 
   [Required(ErrorMessage = "FirstName is required.")]
   public object FirstName{get;set;}
   [Required(ErrorMessage = "LastName is required.")]
   public object LastName{get;set;}
 }
```
Now in a way that is mentioned in [the video](https://www.youtube.com/watch?v=3YwwdoRg2F4)  you have generate the following model and instead of using the Person model you use the Person_Generated(with your own naming convensions) in your whole application :
```

public partial class Person_Generated 
{
   public int Id {get;set;}
   [Required(ErrorMessage = "FirstName is required.")]
   public string FirstName{get;set;}
   [Required(ErrorMessage = "LastName is required.")]
   public string LastName{get;set;}
}
```

you also need to add the following Partial class manually (or maybe you can manage to generate it too for an existing project):

```
public partial class Person_Generated 
{
[NoMap]
  public string FullName => FirstName+" "+LastName;
}

```
another option is also to include the FullName in the first Person_Generated.cs  partial file.

**Comment by Xaeco** [comment]
> @Xaeco Maybe this can help you for now [C# new source generator](https://www.youtube.com/watch?v=3YwwdoRg2F4)

Thanks for your time. What I'm attempting to do is decorate classes that have been auto generated by .proto file compilation so that the DataAnnotationsValidator in Blazor has something to work with.

As per your example I'd need to write a source generator to make a new set of classes based on the auto-generated ones and weave in the data annotations.

**Comment by IanShaughnessyPG** [comment]
I am also very interested in this feature. We would like to decorate classes generated from .proto files as well and use them with the `DataAnnotationsValidator` component.

I dug into the code a bit and it looks like the .NET 5 version of `System.ComponentModel.DataAnnotations.ValidationAttributeStore` is using `System.Reflection` to get the list of attributes, whereas the .NET Framework version is using `TypeDescriptor` and `PropertyDescriptor`.

**Comment by Xaeco** [comment]
I'm going to have a go at fixing it

**Comment by michaelr-sportsbet** [comment]
This seems to be generally the problem with decorating any autogenerated class with attributes. I am experiencing a similar issue with trying to add XmlElement and XmlAttribute to classes generated from a proto file that need to be deserialized from an xml payload. The MetadataType attribute seems to have no effect whatsoever.

I also tried to use MetadataType on a simple class and it seems to be just completely ignored, so it's not something to do with autogenerated files or protobuf, it just seems to not work at all even in the simplest scenario.

(I am using .net standard 2.1 and System.ComponentModel.Annotations nuGet version 5.0.0)

**Comment by Xaeco** [comment]
I created a small test project to investigate the DataAnnotations behaviour. 
Attached zip has two projects with identical code. A .Net Framework project called DN and a .Net Core project called Core.
[TestMetadataTypeAttribute.zip](https://github.com/dotnet/runtime/files/6034298/TestMetadataTypeAttribute.zip)

Some findings:
To get the Metadata type to work in .Net Framework 4.5+ you need to ensure two things
-  When using `Validator.TryValidateObject()` you need to set the `validateAllProperties` parameter to **True**. Otherwise only the `Required` attribute gets checked.
- You need to ensure you add a provider transport as shown in previous posts
``` C#
TypeDescriptor.AddProviderTransparent(
                new AssociatedMetadataTypeTypeDescriptionProvider(typeof(Person), typeof(PersonMetadata)),
                typeof(Person));
```
You don't need to use 
``` c#
[MetadataType(typeof(PersonMetadata))]
```
As it makes no difference to .Net Framework 4.5+

The Core project doesn't produce the expected results.

I went through the .Net Core 5 code, specifically through the System.ComponentModel.Annotations solution. I couldn't find any direct references to MetadataTypeAttribute or AssociatedMetadataType in the context of the TryValidateObject code.

I will keep digging.

**Comment by Xaeco** [comment]
So I have coded two working solutions. 

Solution 1 
- produces the same results as .Net 4.8
- requires `TypeDescriptor.AddProviderTransparent()` method.
- doesn't use the `MetadataType` attribute.
- could potentially be made so that any declaration of the `MetadataType` attribute would auto register the `TypeDescriptor.AddProviderTransparent()` method negating the need for code.

Solution 2 
- produces the same results as .Net 4.8
- uses the `MetadataType` attribute
- doesn't use the `TypeDescriptor.AddProviderTransparent()` method.

Both solutions require adding a couple of methods to the System.ComponentModel.Annotations.Validator class. This resides in .Net Standard 2.1. I'm guessing any fixes would come out in a .NET 6 beta and not to any older versions?

So I'm hoping for some assistance here on which implementation to use. Like to hear everyone's vote?

**Comment by wolfgangschneider** [comment]
I prefer Solution 2 , Its the same as old silverlight WCF RIA services used.

**Comment by georgeemr** [comment]
I'm working on a project on Blazor, and I came through this issue with my current scaffolded models. I hope this gets fixed soon.

**Comment by georgeemr** [comment]
Solution 2 for sure!

**Comment by Xaeco** [comment]
Thank you to all who have commented. 
It looks like solution 2 is preferred. 
I hope to get a pull request in today.

**Comment by Xaeco** [comment]
@SteveSandersonMS I have two working solutions for this issue. 

**Solution 1.** Is to replace two functions with their DN4.8 equivalents. They use TypeDescriptor, which as far as I can tell, is being avoided. 

**Solution 2.** I wrote functions that work but when I do a full build I get an IL Linker error which I can't suppress and I'm not sure I should. 

I'd appreciate talking to someone re the best way forward.

**Comment by SteveSandersonMS** [comment]
@Xaeco Thanks for looking into this.

@eerhardt Do you know who would be a good person on the runtime team to direct this to for triaging/feedback?

**Comment by pranavkm** [comment]
FYI @ajcvickers

**Comment by eerhardt** [comment]
The [listed owners](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) of the `area-System.ComponentModel.DataAnnotations` area are @lajones @ajcvickers.

> Solution 2. I wrote functions that work but when I do a full build I get an IL Linker error which I can't suppress and I'm not sure I should.

@Xaeco - can you try rebasing your change on the latest main? I merged #49901 last week, which resolved the existing ILLink warnings in that library. If you are still seeing warnings, can you send them to me, along with your change? I can assist in getting them resolved.

**Comment by Xaeco** [comment]
Excellent. I will give it a go and let you know. Thank you> The [listed owners](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) of the `area-System.ComponentModel.DataAnnotations` area are @lajones @ajcvickers.
> 
> 
> 
> > Solution 2. I wrote functions that work but when I do a full build I get an IL Linker error which I can't suppress and I'm not sure I should.
> 
> 
> 
> @Xaeco - can you try rebasing your change on the latest main? I merged #49901 last week, which resolved the existing ILLink warnings in that library. If you are still seeing warnings, can you send them to me, along with your change? I can assist in getting them resolved.

**Comment by Xaeco** [comment]
> The [listed owners](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) of the `area-System.ComponentModel.DataAnnotations` area are @lajones @ajcvickers.
> 
> > Solution 2. I wrote functions that work but when I do a full build I get an IL Linker error which I can't suppress and I'm not sure I should.
> 
> @Xaeco - can you try rebasing your change on the latest main? I merged #49901 last week, which resolved the existing ILLink warnings in that library. If you are still seeing warnings, can you send them to me, along with your change? I can assist in getting them resolved.

I've got a PR in for the fix but haven't had anyone look at it yet. Not sure if I've missed part of the process or everyone is really busy and this is too low a priority. Any assistance would be appreciated. Thanks.

**Comment by hdv212** [comment]
Hi all!
What's is the best practice for this situation?
Is there any alternative approach?
Thanks in advance

**Comment by rahul7720** [comment]
@Xaeco Thanks for your contribution. I'm waiting for this fix for a long time.
I'm not sure about the process.  Is there any alternative way to publish the NuGet package or to distribute the dlls?

**Comment by hdv212** [comment]
Hi again.
I'm looking for a best way to accomplish form validation via extending my entities (via dataAnnotantions or someThing else) more than 1 year!
Please give an article or reference to do this via blazor best practice.

**Comment by Xaeco** [comment]
> @Xaeco Thanks for your contribution. I'm waiting for this fix for a long time.
> I'm not sure about the process. Is there any alternative way to publish the NuGet package or to distribute the dlls?
> 
> Will this be released under DotNet 5.0 or DotNet 6.x preview?

Hi. At the moment its in the DotNet 6.x preview codebase. Hopefully it gets approved soon. Meanwhile I'll look into the process to backport it to DotNet 5.0 if that's at all possible.

Currently I am testing my fix locally by incorporating the updated DotNet classes (with a few references changed) into my Blazor project and using a custom validator. The custom validator is required because Blazor's validator is using DotNet 6 Preview internally and that doesn't have my code in it. This works for DotNet 5 & 6.

I'm not sure what the rules are on sharing modified versions of the DotNet files. @eerhardt can you advise?

**Comment by eerhardt** [comment]
> I'm not sure what the rules are on sharing modified versions of the DotNet files. @eerhardt can you advise?

Until the change is merged into `main`, my recommendation would be to pull the change locally and build the file locally.

**Comment by rahul7720** [comment]
Hi Guys,

Any updates on this?

**Comment by Xaeco** [comment]
Still waiting on my pull request to be reviewed. No further progress.

**Comment by Xaeco** [comment]
**Update**. Option 1 has been approved. Not sure if it will land in a .NET 6 preview or the final release.

**Comment by Xaeco** [comment]
@ajcvickers any chance this fix could be back-ported to .NET 5? 

If so what is the process?

**Comment by rahul7720** [comment]
@Xaeco Any chance if you can build an unofficial NuGet package / Dll and release it for testing purposes?

**Comment by bricelam** [comment]
See also (and upvote) https://github.com/dotnet/efcore/issues/14544
